core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -171,6 +178,51 @@ public Expression rewriteComparisonExpression(ComparisonExpression node, Void co
             return unwrapCast(expression);
         }
 
+        @Override
+        public Expression rewriteBetweenPredicate(BetweenPredicate node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            BetweenPredicate expression = (BetweenPredicate) treeRewriter.defaultRewrite((Expression) node, null);
+            return unwrapCast(expression);
+        }
+
+        private Expression unwrapCast(BetweenPredicate expression)
+        {
+            // Canonicalization is handled by CanonicalizeExpressionRewriter
+            if (!(expression.getValue() instanceof Cast cast)) {
+                return expression;
+            }
+
+            Object min = new ExpressionInterpreter(expression.getMin(), plannerContext, session, typeAnalyzer.getTypes(session, types, expression.getMin()))
+                    .optimize(NoOpSymbolResolver.INSTANCE);
+            Object max = new ExpressionInterpreter(expression.getMax(), plannerContext, session, typeAnalyzer.getTypes(session, types, expression.getMax()))
+                    .optimize(NoOpSymbolResolver.INSTANCE);
+
+            if (min == null || min instanceof NullLiteral || max == null || max instanceof NullLiteral) {
+                return new Cast(new NullLiteral(), toSqlType(BOOLEAN));
+            }
+
+            if (min instanceof Expression || max instanceof Expression) {
+                return expression;
+            }
+
+            Type sourceType = typeAnalyzer.getType(session, types, cast.getExpression());
+            Type minType = typeAnalyzer.getType(session, types, expression.getMin());
+            Type maxType = typeAnalyzer.getType(session, types, expression.getMax());
+            verify(minType.equals(maxType), "Mismatched types: %s and %s", minType, maxType);
+
+            if (sourceType instanceof TimestampType && minType == DATE) {
```

Comment:
These conditions aren't necessary

1. decompose between like in `io.trino.sql.planner.DomainTranslator.Visitor#visitBetweenPredicate`
2. delegate to `unwrapCast(ComparisonExpression)`